pkg/providers/instancetype/instancetype.go

```diff
@@ -252,6 +264,58 @@ func (p *DefaultProvider) UpdateInstanceTypeOfferings(ctx context.Context) error
 	return nil
 }
 
+func (p *DefaultProvider) UpdateInstanceTypeMemoryOverhead(ctx context.Context, kubeClient client.Client) error {
+	nodeClassList := &v1.EC2NodeClassList{}
+	if err := kubeClient.List(ctx, nodeClassList); err != nil {
+		return fmt.Errorf("failed to list nodeclasses: %w", err)
+	}
+
+	nodeClassMap := lo.Associate(nodeClassList.Items, func(nodeClass v1.EC2NodeClass) (string, v1.EC2NodeClass) {
+		return nodeClass.Hash(), nodeClass
+	})
+
+	nodeList := &corev1.NodeList{}
+	if err := kubeClient.List(ctx, nodeList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{karpv1.NodeRegisteredLabelKey: "true"}),
+	}); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	instanceTypeInfoMap := lo.Associate(p.instanceTypesInfo, func(i *ec2.InstanceTypeInfo) (string, *ec2.InstanceTypeInfo) {
+		return *i.InstanceType, i
+	})
+
+	workqueue.ParallelizeUntil(ctx, 100, len(nodeList.Items), func(i int) {
+		node := nodeList.Items[i]
+
+		// Get the EC2NodeClass for the current node based on the hash annotation
+		nodeClass, ok := nodeClassMap[node.Annotations[v1.AnnotationEC2NodeClassHash]]
+		if !ok {
+			return
+		}
+
+		instanceType, ok := node.Labels[corev1.LabelInstanceTypeStable]
+		if !ok {
+			return
+		}
+
+		instanceTypeInfo, ok := instanceTypeInfoMap[instanceType]
+		if !ok || instanceTypeInfo == nil {
+			return
+		}
+
+		amiHash, _ := hashstructure.Hash(nodeClass.Status.AMIs, hashstructure.FormatV2, nil)
+
+		// Calculate memory overhead and store to the map
+		reportedMiB := instanceTypeInfo.MemoryInfo.SizeInMiB
+		actualMib := node.Status.Capacity.Memory().Value() / 1024 / 1024
+		overhead := *reportedMiB - actualMib
+		key := fmt.Sprintf("%s-%d", instanceType, amiHash)
+		p.vmMemoryOverheadCache.SetDefault(key, overhead)
```

Comment:
So we assume that last write wins if somehow the allocatables are different?

Reply:
Yes, in theory we shouldn't see different values for the same instance-type+ami pair but if we do we always want the latest.

Reply:
Updated with a check against existing cache value and only updates if equal to or higher than current value. If non-existent will always write.